lightning-graph-sync/src/processing.rs

```diff
@@ -0,0 +1,390 @@
+    chain_source: &Option<CS>,
+) -> Result<(), GraphSyncError>
+where
+    CS::Target: chain::Access,
+{
+    let mut prefix = Vec::with_capacity(4);
+    prefix.resize(4, 0);
+    read_cursor.read_exact(&mut prefix)?;
+
+    let _deserialize_unsigned = match &prefix[..] {
+        &[76, 68, 75, 2] => true,
+        _ => {
+            return Err(GraphSyncError::ProcessingError(
+                "Prefix must equal 'LDK' and the byte 2".to_string(),
+            ));
+        }
+    };
+
+    // shadow variable for ownership release after length-restricted reads
+    let mut read_cursor = read_cursor;
+
+	let announcement_count: BigSize = Readable::read(read_cursor)?;
+	for _ in 0..announcement_count.0 {
+		let message_length: BigSize = Readable::read(read_cursor)?;
+		let mut restricted_cursor = read_cursor.take(message_length.0);
+		let channel_announcement: msgs::UnsignedChannelAnnouncement =
+			Readable::read(&mut restricted_cursor)?;
+		let announcement_result = network_graph.update_channel_from_unsigned_announcement(&channel_announcement, &chain_source);
+		if let Err(lightning_error) = announcement_result {
+			if let ErrorAction::IgnoreDuplicateGossip = lightning_error.action {
+				// everything is fine, just a duplicate channel announcement
+			} else {
+				return Err(lightning_error.into())
+			}
+		}
+
+		read_cursor = restricted_cursor.into_inner();
+	}
+
+	let current_timestamp = SystemTime::now().duration_since(UNIX_EPOCH).unwrap().as_secs() as u32;
+	let update_count: BigSize = Readable::read(read_cursor)?;
+	for update_index in 0..update_count.0 {
+
+		let short_channel_id: u64 = Readable::read(read_cursor)?;
+		let channel_flags: u8 = Readable::read(read_cursor)?;
+
+		if channel_flags & 0b_1000_0000 == 0 {
+			// full update
+			// println!("channel flags: {:b}, full update", channel_flags);
+			let remaining_message_length: BigSize = Readable::read(read_cursor)?;
+
+			let mut restricted_cursor = read_cursor.take(remaining_message_length.0);
+
+			let real_channel_flags = channel_flags & 0b_0000_0011;
+			let message_flags: u8 = Readable::read(&mut restricted_cursor)?;
+
+			let has_htlc_maximum_msat = (message_flags as i32 & 1) == 1;
+
+			let cltv_expiry_delta: u16 = Readable::read(&mut restricted_cursor)?;
+			let htlc_minimum_msat: u64 = Readable::read(&mut restricted_cursor)?;
+			let fee_base_msat: u32 = Readable::read(&mut restricted_cursor)?;
+			let fee_proportional_millionths: u32 = Readable::read(&mut restricted_cursor)?;
+			let htlc_maximum_msat: OptionalField<u64> = if has_htlc_maximum_msat {
+				OptionalField::Present(Readable::read(&mut restricted_cursor).unwrap())
+			} else {
+				OptionalField::Absent
+			};
+
+			let mut excess_data = Vec::new();
+			restricted_cursor.read_to_end(&mut excess_data)?;
+			read_cursor = restricted_cursor.into_inner();
+
+			let synthetic_update = UnsignedChannelUpdate {
+				chain_hash: Default::default(),
+				short_channel_id,
+				timestamp: current_timestamp,
```

Comment:
Hmmmm, I think we need to use the update's real timestamp as otherwise we may get an update from an HTLC failure or otherwise and reject it cause the timestamp is "old".

Reply:
I will get back to this one in a separate commit.

Reply:
Isn't the timestamp only tangentially related to the time of the actual update? I thought nodes had some logic to update the timestamp of a particular update when gossiping anyway – or is the timestamp part of the signed data when a signature is present?

Reply:
Additionally, if the timestamp inaccuracy has implications on HTLC handling, wouldn't that mean that gossip data is slightly more than just a DDoS risk vector?